orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,703 @@
+            frame_layout.addWidget(pp_module_widget)
+        self.store_pipeline()  # Store the pipeline after loading it.
 
-        gui.button(self.controlArea, self, "&Apply", callback=self.apply, default=True)
+        frame_layout.addStretch()
+        self.mainArea.layout().addWidget(frame)
 
-        self.apply()
+    def set_data(self, data=None):
+        self.corpus = data
+        self.update_info()
+        self.commit()
 
-    def select_transformation(self, n):
-        if n > 0:
-            self.transformation = self.transformation_opt[n]
+    def update_info(self):
+        if self.corpus is not None:
+            info = 'Document count: {}'.format(len(self.corpus))
         else:
-            self.transformation = None
-
-    def fill_transformation_options(self):
-        self.trans_combo.clear()
-        for trans in self.transformation_opt:
-            if trans == "(none)":
-                self.trans_combo.addItem("(none)")
-            else:
-                self.trans_combo.addItem(trans)
-
-    def apply(self):
-        # TODO change this to custom stopwords
-        if self.remove_stpwrds:
-            sw = 'english'
-        pp = Preprocessor(incl_punct=self.include_punctuation, trans=self.transformation,
-                          lowercase=self.lowercase, stop_words=sw)
-        self.send(Output.PREPROCESSOR, pp)
\ No newline at end of file
+            info = 'No input corpus detected.'
+        self.info_label.setText(info)
+
+    def commit(self):
+        self.store_pipeline()  # Store the new pipeline.
+        if self.corpus is not None:
+            pp = self.assemble_preprocessor()
+            if pp is not None:
+                self.apply(pp)
+
+    def apply(self, preprocessor):
+        output = preprocessor(self.corpus)
+        self.send(Output.PP_CORPUS, output)
+
+    def assemble_preprocessor(self):
+        self.error(0, '')
+
+        pp_settings = {}
+        for pp in self.preprocessors:
+            if pp.enabled:
+                pp_settings.update(pp.get_pp_setting())
+        try:
+            preprocessor = Preprocessor(**pp_settings)
+        except Exception as e:
+            self.error(0, str(e))
+            return None
+        return preprocessor
+
+    def store_pipeline(self):
+        for pp in self.preprocessors:
+            if isinstance(pp, TokenizerModule):
+                self.persistent_data_tokenizer = pp.export_data()
+            elif isinstance(pp, CasingModule):
+                self.persistent_data_casing = pp.export_data()
+            elif isinstance(pp, TransformationModule):
+                self.persistent_data_stemmer = pp.export_data()
+            elif isinstance(pp, FilteringModule):
+                self.persistent_data_filter = pp.export_data()
+
+    @Slot()
+    def settings_invalidated(self):
+        self.commit()
+
+    @Slot(str)
+    def display_message(self, message):
+        self.error(0, message)
```

Comment:
Each widget module should have a 
```py
if __name__ == '__main__':
    ...  # Construct the app and show the widget with example data
```
For an example, have a look at [any widget from orange](https://github.com/biolab/orange3/).